{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "< # of observed mutations >\n",
      "504\n",
      "\n",
      "< # of unique mutations >\n",
      "202\n",
      "\n",
      "< # of genes >\n",
      "34\n",
      "\n",
      "< # of mutations in tissue >\n",
      "232\n",
      "\n",
      "< # of mutations in baseline >\n",
      "155\n",
      "\n",
      "< # of mutations in PD >\n",
      "117\n",
      "\n",
      "< Variant classifications >\n",
      "Missense_Mutation    272\n",
      "Nonsense_Mutation    101\n",
      "Frame_Shift_Del       68\n",
      "Frame_Shift_Ins       50\n",
      "Splice_Site           13\n",
      "Name: Variant_Classification, dtype: int64\n",
      "\n",
      "< Samples with no mutations at all >\n",
      "['CRC168']\n",
      "\n",
      "< Mutational prevalence in APC >\n",
      "Tissue: 67.9\n",
      "Baseline: 44.0\n",
      "PD: 34.5\n",
      "\n",
      "< # of samples who benifited by sequencing baseline-ctDNA >\n",
      "32 38.1\n",
      "\n",
      "< # of samples who benifited by sequencing both baseline- and PD-ctDNA >\n",
      "40 47.6\n",
      "\n",
      "< # of patients who did not have any mutations in tDNA >\n",
      "4 4.8\n",
      "\n",
      "< # of patients who did not have any mutations in baseline-ctDNA >\n",
      "22 26.2\n",
      "\n",
      "< # of patients who did not have any mutations in PD-ctDNA >\n",
      "33 39.3\n",
      "\n",
      "< # of subcloncal mutations >\n",
      "20 4.0\n"
     ]
    }
   ],
   "source": [
    "from fuc import pymaf, common\n",
    "\n",
    "mf = pymaf.MafFrame.from_file('./data/final.maf')\n",
    "mf.df['PatientID'] = [x.split('_')[0] for x in mf.df.Tumor_Sample_Barcode]\n",
    "mf.df['SampleType'] = [x.split('_')[1] for x in mf.df.Tumor_Sample_Barcode]\n",
    "mf.df['Clonality'] = mf.compute_clonality('AF')\n",
    "\n",
    "af = common.AnnFrame.from_file('./data/sample-annot.csv', sep=',', sample_col='SampleID')\n",
    "\n",
    "n_patients = af.df.StudyID.nunique()\n",
    "n_mutations = mf.shape[0]\n",
    "\n",
    "t_df = mf.df[mf.df.SampleType == 'T']\n",
    "b_df = mf.df[mf.df.SampleType == 'B']\n",
    "p_df = mf.df[mf.df.SampleType == 'P']\n",
    "\n",
    "print('< # of observed mutations >')\n",
    "print(mf.shape[0])\n",
    "print()\n",
    "print('< # of unique mutations >')\n",
    "print(mf.df[~mf.df.duplicated(['Hugo_Symbol', 'Protein_Change'])].shape[0])\n",
    "print()\n",
    "print('< # of genes >')\n",
    "print(mf.df.Hugo_Symbol.nunique())\n",
    "print()\n",
    "print('< # of mutations in tissue >')\n",
    "print(t_df.shape[0])\n",
    "print()\n",
    "print('< # of mutations in baseline >')\n",
    "print(b_df.shape[0])\n",
    "print()\n",
    "print('< # of mutations in PD >')\n",
    "print(p_df.shape[0])\n",
    "print()\n",
    "print('< Variant classifications >')\n",
    "print(mf.df.Variant_Classification.value_counts())\n",
    "print()\n",
    "print('< Samples with no mutations at all >')\n",
    "print([x for x in af.df.StudyID.unique() if x not in [x.split('_')[0] for x in mf.df.Tumor_Sample_Barcode]])\n",
    "print()\n",
    "print('< Mutational prevalence in APC >')\n",
    "apc_df = mf.df[mf.df.Hugo_Symbol == 'APC']\n",
    "print('Tissue:', round(apc_df[apc_df.SampleType == 'T'].PatientID.nunique() / af.df.StudyID.nunique() * 100, 1))\n",
    "print('Baseline:', round(apc_df[apc_df.SampleType == 'B'].PatientID.nunique() / af.df.StudyID.nunique() * 100, 1))\n",
    "print('PD:', round(apc_df[apc_df.SampleType == 'P'].PatientID.nunique() / af.df.StudyID.nunique() * 100, 1))\n",
    "print()\n",
    "print('< # of samples who benifited by sequencing baseline-ctDNA >')\n",
    "patients = []\n",
    "for patient in af.df.StudyID.unique():\n",
    "    for gene in mf.df.Hugo_Symbol.unique():\n",
    "        df = mf.df[(mf.df.PatientID == patient) & (mf.df.Hugo_Symbol == gene)]\n",
    "        for mutation in df.Protein_Change.unique():\n",
    "            l = df[df.Protein_Change == mutation].SampleType.to_list()\n",
    "            if 'T' not in l and 'B' in l:\n",
    "                patients.append(patient)\n",
    "print(len(set(patients)), round(len(set(patients)) / n_patients * 100, 1))\n",
    "print()\n",
    "print('< # of samples who benifited by sequencing both baseline- and PD-ctDNA >')\n",
    "patients = []\n",
    "for patient in af.df.StudyID.unique():\n",
    "    for gene in mf.df.Hugo_Symbol.unique():\n",
    "        df = mf.df[(mf.df.PatientID == patient) & (mf.df.Hugo_Symbol == gene)]\n",
    "        for mutation in df.Protein_Change.unique():\n",
    "            l = df[df.Protein_Change == mutation].SampleType.to_list()\n",
    "            if 'T' not in l and ('B' in l or 'P' in l):\n",
    "                patients.append(patient)\n",
    "print(len(set(patients)), round(len(set(patients)) / n_patients * 100, 1))\n",
    "print()\n",
    "print('< # of patients who did not have any mutations in tDNA >')\n",
    "print(n_patients - t_df.PatientID.nunique(), round((n_patients - t_df.PatientID.nunique()) / n_patients * 100, 1))\n",
    "print()\n",
    "print('< # of patients who did not have any mutations in baseline-ctDNA >')\n",
    "print(n_patients - b_df.PatientID.nunique(), round((n_patients - b_df.PatientID.nunique()) / n_patients * 100, 1))\n",
    "print()\n",
    "print('< # of patients who did not have any mutations in PD-ctDNA >')\n",
    "print(n_patients - p_df.PatientID.nunique(), round((n_patients - p_df.PatientID.nunique()) / n_patients * 100, 1))\n",
    "print()\n",
    "print('< # of subcloncal mutations >')\n",
    "print(mf.df.Clonality.value_counts()['Subclonal'] , round(mf.df.Clonality.value_counts()['Subclonal'] / n_mutations * 100, 1))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Figure 1. Overview of sample collection scheme and data analysis pipeline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![Overview](overview.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Figure 2. Oncoplot of genomic alterations in colorectal cancer patients for the ten most commonly mutated genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from fuc import pymaf, common\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "\n",
    "mf = pymaf.MafFrame.from_file('./data/final.maf')\n",
    "af1 = common.AnnFrame.from_file('./data/sample-annot.csv', sep=',', sample_col='SampleID')\n",
    "af2 = common.AnnFrame.from_file('./data/patient-annot.csv', sep=',', sample_col='StudyID')\n",
    "\n",
    "figsize = (15, 10)\n",
    "label_fontsize = 12\n",
    "ticklabels_fontsize = 12\n",
    "legend_fontsize = 11\n",
    "patient_col = 'StudyID'\n",
    "group_col = 'Type'\n",
    "groups = ['Tissue', 'Baseline', 'PD']\n",
    "colors = 'Set2'\n",
    "seq_colors = ['yellow', 'limegreen', 'darkgreen', 'navy']\n",
    "\n",
    "fig, axes = plt.subplots(8, 2, figsize=figsize, \n",
    "                         gridspec_kw={'height_ratios': [1, 10, 1/3, 1/3, 1/3, 1/3, 1/3, 4], \n",
    "                                      'width_ratios': [10, 1.5]})\n",
    "\n",
    "[[ax1, ax2], [ax3, ax4], [ax5, ax6], [ax7, ax8], \n",
    " [ax9, ax10], [ax11, ax12], [ax13, ax14], [ax15, ax16]] = axes\n",
    "\n",
    "patients = mf.matrix_waterfall_matched(af1, patient_col, group_col, groups).columns\n",
    "\n",
    "\n",
    "mf.plot_tmb_matched(af1, patient_col, group_col, group_order=groups, ax=ax1, legend=False, \n",
    "                    patients=patients, width=0.8, color=sns.color_palette(colors)[:3])\n",
    "ax1.set_xticks([])\n",
    "ax1.set_xlim(-0.5, len(patients)-0.5)\n",
    "ax1.spines['right'].set_visible(False)\n",
    "ax1.spines['top'].set_visible(False)\n",
    "ax1.spines['bottom'].set_visible(False)\n",
    "ax1.set_ylabel('TMB', fontsize=label_fontsize)\n",
    "ax1.tick_params(axis='y', which='major',\n",
    "                labelsize=ticklabels_fontsize)\n",
    "\n",
    "ax2.remove()\n",
    "\n",
    "mf.plot_waterfall_matched(af1, patient_col, group_col, group_order=groups, ax=ax3)\n",
    "ax3.set_xticks([])\n",
    "ax3.tick_params(axis='y', which='major', labelrotation=0,\n",
    "                labelsize=ticklabels_fontsize)\n",
    "\n",
    "mf.plot_mutated_matched(\n",
    "    af1, patient_col, group_col, group_order=groups, ax=ax4, palette=colors\n",
    ")\n",
    "ax4.set_yticks([])\n",
    "ax4.legend().remove()\n",
    "ax4.spines['right'].set_visible(False)\n",
    "ax4.spines['left'].set_visible(False)\n",
    "ax4.spines['top'].set_visible(False)\n",
    "ax4.tick_params(axis='x', which='major',\n",
    "                labelsize=ticklabels_fontsize)\n",
    "ax4.set_xlabel('Patients (%)', fontsize=label_fontsize)\n",
    "\n",
    "ax5, handles3 = af2.plot_annot(\n",
    "    'Panel', \n",
    "    group_order=['SNUBH_V1', 'SNUBH_V2', 'TS500', 'ACMP'],\n",
    "    ax=ax5,\n",
    "    samples=patients,\n",
    "    xticklabels=False,\n",
    "    colors=list(plt.get_cmap('Set2').colors[3:7])\n",
    ")\n",
    "ax5.set_ylabel('Panel', rotation='horizontal', ha='right', va='center', fontsize=12)\n",
    "\n",
    "ax6.remove()\n",
    "\n",
    "ax7, handles4 = af2.plot_annot(\n",
    "    'MaxVAF',\n",
    "    ax=ax7,\n",
    "    samples=patients,\n",
    "    group_order=['VAF < 0.2', '0.2 <= VAF < 0.4', '0.4 <= VAF < 0.6', 'VAF > 0.6'],\n",
    "    colors=seq_colors,\n",
    "    xticklabels=False,\n",
    "    sequential=True\n",
    ")\n",
    "ax7.set_ylabel('Max VAF', rotation='horizontal', ha='right', va='center', fontsize=12)\n",
    "\n",
    "ax8.remove()\n",
    "\n",
    "ax9, handles6 = af2.plot_annot(\n",
    "    'AgeGroup',\n",
    "    ax=ax9,\n",
    "    samples=patients,\n",
    "    group_order=['<50', '50s', '60s', '>=70'],\n",
    "    colors=seq_colors,\n",
    "    xticklabels=False,\n",
    "    sequential=True\n",
    ")\n",
    "ax9.set_ylabel('Age', rotation='horizontal', ha='right', va='center', fontsize=12)\n",
    "\n",
    "ax10.remove()\n",
    "\n",
    "ax11, handles5 = af2.plot_annot(\n",
    "    'Sex',\n",
    "    ax=ax11,\n",
    "    samples=patients,\n",
    "    group_order=['Female', 'Male'],\n",
    "    xticklabels=False,\n",
    "    colors=['red', 'blue']\n",
    ")\n",
    "ax11.set_ylabel('Sex', rotation='horizontal', ha='right', va='center', fontsize=12)\n",
    "\n",
    "ax12.remove()\n",
    "\n",
    "ax13, handles7 = af2.plot_annot(\n",
    "    'MSI',\n",
    "    ax=ax13,\n",
    "    samples=patients,\n",
    "    group_order=['MSS', 'MSI-L'],\n",
    "    colors=['red', 'blue']\n",
    ")\n",
    "ax13.set_ylabel('MSI', rotation='horizontal', ha='right', va='center', fontsize=12)\n",
    "\n",
    "ax14.remove()\n",
    "\n",
    "# # Create the legends.\n",
    "kwargs = dict(fontsize=legend_fontsize, title_fontsize=legend_fontsize)\n",
    "handles1 = common.legend_handles(pymaf.NONSYN_NAMES+['Multi_Hit'], colors=pymaf.NONSYN_COLORS+['k'])\n",
    "handles2 = common.legend_handles(groups, colors=colors)\n",
    "leg1 = ax15.legend(handles=handles1, loc=(0, 0), title='Variant Classification', ncol=2, **kwargs)\n",
    "leg2 = ax15.legend(handles=handles2, loc=(0.375, 0), title=group_col, **kwargs)\n",
    "leg3 = ax15.legend(handles=handles3, loc=(0.485, 0), title='Panel', **kwargs)\n",
    "leg4 = ax15.legend(handles=handles4, loc=(0.608, 0), title='Max VAF in Tissue', **kwargs)\n",
    "leg5 = ax15.legend(handles=handles6, loc=(0.775, 0), title='Age', **kwargs)\n",
    "leg6 = ax15.legend(handles=handles5, loc=(0.866, 0), title='Sex', **kwargs)\n",
    "leg7 = ax15.legend(handles=handles7, loc=(0.966, 0), title='MSI', **kwargs)\n",
    "\n",
    "ax15.add_artist(leg1)\n",
    "ax15.add_artist(leg2)\n",
    "ax15.add_artist(leg3)\n",
    "ax15.add_artist(leg4)\n",
    "ax15.add_artist(leg5)\n",
    "ax15.add_artist(leg6)\n",
    "ax15.axis('off')\n",
    "\n",
    "# Remove the bottom right plot.\n",
    "ax16.remove()\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.subplots_adjust(wspace=0.01, hspace=0.01)\n",
    "plt.savefig('./outputs/oncoplot.png')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Figure 3. Correlation analysis results from pairwise comparisons between DNA sample types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from fuc import pymaf, pyvcf, common\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "\n",
    "sns.set()\n",
    "\n",
    "mf = pymaf.MafFrame.from_file('./data/final.maf')\n",
    "af = common.AnnFrame.from_file('./data/sample-annot.csv', sep=',', sample_col='SampleID')\n",
    "\n",
    "fig, axes = plt.subplots(2, 4, figsize=(20, 10))\n",
    "\n",
    "[[ax1, ax2, ax3, ax4], [ax5, ax6, ax7, ax8]] = axes\n",
    "\n",
    "gs = axes[0, 3].get_gridspec()\n",
    "\n",
    "# remove the underlying axes\n",
    "for ax in axes[:, 3]:\n",
    "    ax.remove()\n",
    "    \n",
    "axbig = fig.add_subplot(gs[:, 3])\n",
    "\n",
    "n = 84\n",
    "\n",
    "mf.plot_regplot_gene(af, 'Type', 'Tissue', 'Baseline', ax=ax1, a_size=n, b_size=n, to_csv='./data/correlations-baseline_tissue.csv')\n",
    "mf.plot_regplot_gene(af, 'Type', 'Tissue', 'PD', ax=ax2, a_size=n, b_size=n, to_csv='./data/correlations-tissue_pd.csv')\n",
    "mf.plot_regplot_gene(af, 'Type', 'Baseline', 'PD', ax=ax3, a_size=n, b_size=n, to_csv='./data/correlations-baseline_pd.csv')\n",
    "\n",
    "mf.plot_regplot_tmb(af, 'StudyID', 'Type', 'Tissue', 'Baseline', ax=ax5, to_csv='./data/tmb-baseline_tissue.csv')\n",
    "mf.plot_regplot_tmb(af, 'StudyID', 'Type', 'Tissue', 'PD', ax=ax6, to_csv='./data/tmb-tissue_pd.csv')\n",
    "mf.plot_regplot_tmb(af, 'StudyID', 'Type', 'Baseline', 'PD', ax=ax7, to_csv='./data/tmb-baseline_pd.csv')\n",
    "\n",
    "mf.plot_vaf('AF', group_col='Type', group_order=['Tissue', 'Baseline', 'PD'], af=af, ax=axbig, sort=False, flip=True)\n",
    "\n",
    "for i, ax in enumerate([ax1, ax2, ax3, ax5, ax6, ax7]):\n",
    "    if i < 3:\n",
    "        ax.set_xlim([-0.05, 0.8])\n",
    "        ax.set_ylim([-0.05, 0.8])\n",
    "        ax.set_yticks([0.0, 0.2, 0.4, 0.6, 0.8])\n",
    "    else:\n",
    "        ax.set_xlim([-0.5, 8.5])\n",
    "        ax.set_ylim([-0.5, 8.5])\n",
    "        ax.set_yticks([0, 2, 4, 6, 8])\n",
    "\n",
    "    ax.xaxis.label.set_size(15)\n",
    "    ax.yaxis.label.set_size(15)\n",
    "\n",
    "    ax.tick_params(axis='both', which='major', labelsize=15)\n",
    "\n",
    "axbig.xaxis.label.set_size(15)\n",
    "axbig.tick_params(axis='both', which='major', labelsize=15)\n",
    "axbig.legend(fontsize=15)\n",
    "\n",
    "ax2.set_title('A', fontsize=20)\n",
    "ax6.set_title('B', fontsize=20)\n",
    "axbig.set_title('C', fontsize=20)\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.savefig('./outputs/correlations.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib\n",
    "matplotlib.rc_file_defaults()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Figure 4. Analysis of gene-gene interaction in colorectal cancer patients"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from fuc import pymaf, common\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "\n",
    "mf = pymaf.MafFrame.from_file('./data/final.maf')\n",
    "af = common.AnnFrame.from_file('./data/sample-annot.csv', sep=',', sample_col='SampleID')\n",
    "\n",
    "fig, axes = plt.subplots(2, 3, figsize=(16, 8), gridspec_kw={'width_ratios': [1, 0.5, 0.5]})\n",
    "\n",
    "gs = axes[0, 0].get_gridspec()\n",
    "\n",
    "axes[0, 0].remove()\n",
    "axes[1, 0].remove()\n",
    "axbig = fig.add_subplot(gs[:, 0])\n",
    "\n",
    "mf.plot_interactions(cmap='BrBG', linewidths=0.5, ax=axbig)\n",
    "\n",
    "for ticklabel in axbig.get_yticklabels():\n",
    "    ticklabel.set_rotation(0)\n",
    "\n",
    "mf.plot_genepair('KRAS', 'APC', 'AF', af=af, group_col='Type', ax=axes[0, 1], group_order=['Tissue', 'Baseline', 'PD'], alpha=0.7)\n",
    "mf.plot_genepair('KRAS', 'TP53', 'AF', af=af, group_col='Type', ax=axes[0, 2], group_order=['Tissue', 'Baseline', 'PD'], alpha=0.7)\n",
    "mf.plot_genepair('KRAS', 'BRAF', 'AF', af=af, group_col='Type', ax=axes[1, 1], group_order=['Tissue', 'Baseline', 'PD'], alpha=0.7)\n",
    "mf.plot_genepair('KRAS', 'NRAS', 'AF', af=af, group_col='Type', ax=axes[1, 2], group_order=['Tissue', 'Baseline', 'PD'], alpha=0.7)\n",
    "\n",
    "ax_list = [axbig, axes[0, 1], axes[0, 2], axes[1, 1], axes[1, 2]]\n",
    "titles = ['A', 'B', 'C', 'D', 'E']\n",
    "\n",
    "for i, ax in enumerate(ax_list):\n",
    "    ax.set_title(titles[i], fontsize=20)\n",
    "    if i in [1, 2, 3]:\n",
    "        ax.legend().remove()\n",
    "    if i in [1, 2, 3, 4]:\n",
    "        ax.set_xlim([-0.05, 1.05])\n",
    "        ax.set_ylim([-0.05, 1.05])\n",
    "    ax.xaxis.label.set_size(15)\n",
    "    ax.yaxis.label.set_size(15)\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.savefig('./outputs/interactions.png')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Figure 5. Representative examples of cancer evolution dynamics."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from fuc import pymaf, common\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "\n",
    "sns.set()\n",
    "\n",
    "mf = pymaf.MafFrame.from_file('./data/final.maf')\n",
    "af = common.AnnFrame.from_file('./data/sample-annot.csv', sep=',', sample_col='SampleID')\n",
    "df = af.df.reset_index().pivot(index='StudyID', columns='Type', values='SampleID').reset_index()\n",
    "d = {}\n",
    "\n",
    "for i, r in df.iterrows():\n",
    "    d[r.StudyID] = [r.Tissue, r.Baseline, r.PD]\n",
    "\n",
    "fig, [[ax1, ax2, ax3], [ax4, ax5, ax6]] = plt.subplots(2, 3, figsize=(20, 15))\n",
    "\n",
    "axes = [ax1, ax2, ax3, ax4, ax5, ax6]\n",
    "patients = ['CRC105', 'CRC035', 'CRC080', 'CRC019', 'CRC036', 'CRC065']\n",
    "\n",
    "for i, ax in enumerate(axes):\n",
    "    mf.plot_evolution(d[patients[i]], 'AF', ax=ax, count=50)\n",
    "    ax.set_ylim([-0.1, 1.1])\n",
    "    ax.set_xticklabels(['Tissue', 'Baseline', 'PD'])\n",
    "    ax.tick_params(axis='both', which='major', labelsize=15)\n",
    "    ax.yaxis.label.set_size(20)\n",
    "    ax.set_title(patients[i], fontsize=20)\n",
    "    ax.legend(fontsize=15)\n",
    "    \n",
    "plt.tight_layout()\n",
    "plt.savefig('./outputs/evolution.png')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Table 1. Gene concordance for the ten genes with the highest mutation prevalence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from fuc import pymaf, common\n",
    "import pandas as pd\n",
    "\n",
    "mf = pymaf.MafFrame.from_file('./data/final.maf')\n",
    "af = common.AnnFrame.from_file('./data/sample-annot.csv', sep=',', sample_col='SampleID')\n",
    "df = af.df.reset_index().pivot(index='StudyID', columns='Type', values='SampleID')\n",
    "\n",
    "genes = ['APC', 'TP53', 'KRAS', 'SMAD4', 'FBXW7', 'BRAF', 'PTEN', 'NRAS', 'PIK3CA', 'RET']\n",
    "conc = {}\n",
    "\n",
    "for gene in genes:\n",
    "    TB = []\n",
    "    TP = []\n",
    "    BP = []\n",
    "    for i, r in df.iterrows():        \n",
    "        TB.append(mf.get_gene_concordance(gene, r.Tissue, r.Baseline))\n",
    "        TP.append(mf.get_gene_concordance(gene, r.Tissue, r.PD))\n",
    "        BP.append(mf.get_gene_concordance(gene, r.Baseline, r.PD))\n",
    "    conc[gene] = [sum(TB), sum(TP), sum(BP)]\n",
    "    conc[gene] = [round(x/df.shape[0] * 100, 1) for x in conc[gene]]\n",
    "    \n",
    "df = pd.DataFrame(conc).T\n",
    "df.index.name = 'Gene'\n",
    "df.columns = ['TB', 'TP', 'BP']\n",
    "df.to_csv('./outputs/gene-concordance.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Supplementary Figure 1. Lollipop plot for the ten most commonly mutated genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from fuc import pymaf, common\n",
    "import seaborn as sns\n",
    "\n",
    "sns.set()\n",
    "\n",
    "mf = pymaf.MafFrame.from_file('./data/final.maf')\n",
    "af = common.AnnFrame.from_file('./data/sample-annot.csv', sep=',', sample_col='SampleID')\n",
    "\n",
    "t_mf = mf.subset(af.df[af.df.Type == 'Tissue'].index.to_list())\n",
    "b_mf = mf.subset(af.df[af.df.Type == 'Baseline'].index.to_list())\n",
    "p_mf = mf.subset(af.df[af.df.Type == 'PD'].index.to_list())\n",
    "\n",
    "fig, axes = plt.subplots(3, 3, figsize=(20, 15))\n",
    "\n",
    "specimens = ['Tissue', 'Baseline', 'PD']\n",
    "genes = ['APC', 'TP53', 'KRAS']\n",
    "mfs = [t_mf, b_mf, p_mf]\n",
    "xmin = [150, 60, 5]\n",
    "xmax = [1650, 360, 150]\n",
    "ymax = [6.2, 5.2, 12.2]\n",
    "\n",
    "i = 0\n",
    "j = 0\n",
    "\n",
    "for row in axes:\n",
    "    for col in row:\n",
    "        mfs[i].plot_lollipop(genes[j], ax=col)\n",
    "        if i == 0 and j == 2:\n",
    "            col.legend(fontsize=15, markerscale=2)\n",
    "        else:\n",
    "            col.get_legend().remove()\n",
    "        if i == 0:\n",
    "            col.set_title(genes[j], fontsize=20)\n",
    "        if j == 0:\n",
    "            col.set_ylabel(specimens[i])\n",
    "        else:\n",
    "            col.set_ylabel('')\n",
    "        if i != 2:\n",
    "            col.set_xlabel('')\n",
    "        else:\n",
    "            col.xaxis.label.set_size(20)\n",
    "        col.tick_params(axis='both', which='major', labelsize=15)\n",
    "        col.yaxis.label.set_size(20)\n",
    "        col.set_xlim(xmin=xmin[j], xmax=xmax[j])\n",
    "        col.set_ylim(ymax=ymax[j])\n",
    "        j += 1\n",
    "    i += 1\n",
    "    j = 0\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.savefig('./outputs/lollipop.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib\n",
    "matplotlib.rc_file_defaults()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Supplementary Figure 2. Oncoplot of genomic alterations in colorectal cancer patients for all 43 mutated genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from fuc import pymaf, common\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "\n",
    "mf = pymaf.MafFrame.from_file('./data/final.maf')\n",
    "af1 = common.AnnFrame.from_file('./data/sample-annot.csv', sep=',', sample_col='SampleID')\n",
    "af2 = common.AnnFrame.from_file('./data/patient-annot.csv', sep=',', sample_col='StudyID')\n",
    "\n",
    "figsize = (15, 17)\n",
    "label_fontsize = 12\n",
    "ticklabels_fontsize = 12\n",
    "legend_fontsize = 11\n",
    "count=100\n",
    "patient_col = 'StudyID'\n",
    "group_col = 'Type'\n",
    "groups = ['Tissue', 'Baseline', 'PD']\n",
    "colors = 'Set2'\n",
    "seq_colors = ['yellow', 'limegreen', 'darkgreen', 'navy']\n",
    "\n",
    "fig, axes = plt.subplots(8, 2, figsize=figsize, \n",
    "                         gridspec_kw={'height_ratios': [1, 18, 1/4, 1/4, 1/4, 1/4, 1/4, 3.2], \n",
    "                                      'width_ratios': [10, 1.5]})\n",
    "\n",
    "[[ax1, ax2], [ax3, ax4], [ax5, ax6], [ax7, ax8], \n",
    " [ax9, ax10], [ax11, ax12], [ax13, ax14], [ax15, ax16]] = axes\n",
    "\n",
    "patients = mf.matrix_waterfall_matched(af1, patient_col, group_col, groups, count=count).columns\n",
    "\n",
    "\n",
    "mf.plot_tmb_matched(af1, patient_col, group_col, group_order=groups, ax=ax1, legend=False, \n",
    "                    patients=patients, width=0.8, color=sns.color_palette(colors)[:3])\n",
    "ax1.set_xticks([])\n",
    "ax1.set_xlim(-0.5, len(patients)-0.5)\n",
    "ax1.spines['right'].set_visible(False)\n",
    "ax1.spines['top'].set_visible(False)\n",
    "ax1.spines['bottom'].set_visible(False)\n",
    "ax1.set_ylabel('TMB', fontsize=label_fontsize)\n",
    "ax1.tick_params(axis='y', which='major',\n",
    "                labelsize=ticklabels_fontsize)\n",
    "\n",
    "ax2.remove()\n",
    "\n",
    "mf.plot_waterfall_matched(af1, patient_col, group_col, group_order=groups, ax=ax3, count=count)\n",
    "ax3.set_xticks([])\n",
    "ax3.tick_params(axis='y', which='major', labelrotation=0,\n",
    "                labelsize=ticklabels_fontsize)\n",
    "\n",
    "mf.plot_mutated_matched(\n",
    "    af1, patient_col, group_col, group_order=groups, ax=ax4, palette=colors, count=count\n",
    ")\n",
    "ax4.set_yticks([])\n",
    "ax4.legend().remove()\n",
    "ax4.spines['right'].set_visible(False)\n",
    "ax4.spines['left'].set_visible(False)\n",
    "ax4.spines['top'].set_visible(False)\n",
    "ax4.tick_params(axis='x', which='major',\n",
    "                labelsize=ticklabels_fontsize)\n",
    "ax4.set_xlabel('Patients (%)', fontsize=label_fontsize)\n",
    "\n",
    "ax5, handles3 = af2.plot_annot(\n",
    "    'Panel', \n",
    "    group_order=['SNUBH_V1', 'SNUBH_V2', 'TS500', 'ACMP'],\n",
    "    ax=ax5,\n",
    "    samples=patients,\n",
    "    xticklabels=False,\n",
    "    colors=list(plt.get_cmap('Set2').colors[3:7])\n",
    ")\n",
    "ax5.set_ylabel('Panel', rotation='horizontal', ha='right', va='center', fontsize=12)\n",
    "\n",
    "ax6.remove()\n",
    "\n",
    "ax7, handles4 = af2.plot_annot(\n",
    "    'MaxVAF',\n",
    "    ax=ax7,\n",
    "    samples=patients,\n",
    "    group_order=['VAF < 0.2', '0.2 <= VAF < 0.4', '0.4 <= VAF < 0.6', 'VAF > 0.6'],\n",
    "    colors=seq_colors,\n",
    "    xticklabels=False,\n",
    "    sequential=True\n",
    ")\n",
    "ax7.set_ylabel('Max VAF', rotation='horizontal', ha='right', va='center', fontsize=12)\n",
    "\n",
    "ax8.remove()\n",
    "\n",
    "ax9, handles6 = af2.plot_annot(\n",
    "    'AgeGroup',\n",
    "    ax=ax9,\n",
    "    samples=patients,\n",
    "    group_order=['<50', '50s', '60s', '>=70'],\n",
    "    colors=seq_colors,\n",
    "    xticklabels=False,\n",
    "    sequential=True\n",
    ")\n",
    "ax9.set_ylabel('Age', rotation='horizontal', ha='right', va='center', fontsize=12)\n",
    "\n",
    "ax10.remove()\n",
    "\n",
    "ax11, handles5 = af2.plot_annot(\n",
    "    'Sex',\n",
    "    ax=ax11,\n",
    "    samples=patients,\n",
    "    group_order=['Female', 'Male'],\n",
    "    xticklabels=False,\n",
    "    colors=['red', 'blue']\n",
    ")\n",
    "ax11.set_ylabel('Sex', rotation='horizontal', ha='right', va='center', fontsize=12)\n",
    "\n",
    "ax12.remove()\n",
    "\n",
    "ax13, handles7 = af2.plot_annot(\n",
    "    'MSI',\n",
    "    ax=ax13,\n",
    "    samples=patients,\n",
    "    group_order=['MSS', 'MSI-L'],\n",
    "    colors=['red', 'blue']\n",
    ")\n",
    "ax13.set_ylabel('MSI', rotation='horizontal', ha='right', va='center', fontsize=12)\n",
    "\n",
    "ax14.remove()\n",
    "\n",
    "# # Create the legends.\n",
    "kwargs = dict(fontsize=legend_fontsize, title_fontsize=legend_fontsize)\n",
    "handles1 = common.legend_handles(pymaf.NONSYN_NAMES+['Multi_Hit'], colors=pymaf.NONSYN_COLORS+['k'])\n",
    "handles2 = common.legend_handles(groups, colors=colors)\n",
    "leg1 = ax15.legend(handles=handles1, loc=(0, 0), title='Variant Classification', ncol=2, **kwargs)\n",
    "leg2 = ax15.legend(handles=handles2, loc=(0.375, 0), title=group_col, **kwargs)\n",
    "leg3 = ax15.legend(handles=handles3, loc=(0.485, 0), title='Panel', **kwargs)\n",
    "leg4 = ax15.legend(handles=handles4, loc=(0.608, 0), title='Max VAF in Tissue', **kwargs)\n",
    "leg5 = ax15.legend(handles=handles6, loc=(0.775, 0), title='Age', **kwargs)\n",
    "leg6 = ax15.legend(handles=handles5, loc=(0.866, 0), title='Sex', **kwargs)\n",
    "leg7 = ax15.legend(handles=handles7, loc=(0.966, 0), title='MSI', **kwargs)\n",
    "\n",
    "ax15.add_artist(leg1)\n",
    "ax15.add_artist(leg2)\n",
    "ax15.add_artist(leg3)\n",
    "ax15.add_artist(leg4)\n",
    "ax15.add_artist(leg5)\n",
    "ax15.add_artist(leg6)\n",
    "ax15.axis('off')\n",
    "\n",
    "# Remove the bottom right plot.\n",
    "ax16.remove()\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.subplots_adjust(wspace=0.01, hspace=0.01)\n",
    "plt.savefig('./outputs/oncoplot-all.png')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Supplementary Figure 3. Weak gene-gene interactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from fuc import pymaf, common\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "\n",
    "mf = pymaf.MafFrame.from_file('./data/final.maf')\n",
    "af = common.AnnFrame.from_file('./data/sample-annot.csv', sep=',', sample_col='SampleID')\n",
    "\n",
    "fig, [[ax1, ax2], [ax3, ax4]] = plt.subplots(2, 2, figsize=(12, 12))\n",
    "\n",
    "mf.plot_genepair('PTEN', 'BRAF', 'AF', af=af, group_col='Type', ax=ax1, group_order=['Tissue', 'Baseline', 'PD'], alpha=0.7)\n",
    "mf.plot_genepair('PTEN', 'NRAS', 'AF', af=af, group_col='Type', ax=ax2, group_order=['Tissue', 'Baseline', 'PD'], alpha=0.7)\n",
    "mf.plot_genepair('PIK3CA', 'BRAF', 'AF', af=af, group_col='Type', ax=ax3, group_order=['Tissue', 'Baseline', 'PD'], alpha=0.7)\n",
    "mf.plot_genepair('APC', 'RET', 'AF', af=af, group_col='Type', ax=ax4, group_order=['Tissue', 'Baseline', 'PD'], alpha=0.7)\n",
    "\n",
    "for i, ax in enumerate([ax1, ax2, ax3, ax4]):\n",
    "    ax.tick_params(axis='both', which='major', labelsize=15)\n",
    "    ax.xaxis.label.set_size(20)\n",
    "    ax.yaxis.label.set_size(20)\n",
    "    if i == 0:\n",
    "        ax.legend(fontsize=15)\n",
    "    else:\n",
    "        ax.legend().remove()\n",
    "    if i == 3:\n",
    "        ax.set_title('Mutually exclusive', fontsize=20)\n",
    "    else:\n",
    "        ax.set_title('Co-occurring', fontsize=20)\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.savefig('./outputs/interactions-spurious.png')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Supplementary Figure 4. Identification of subclonal mutations using different thresholds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from fuc import pymaf, common\n",
    "import seaborn as sns\n",
    "\n",
    "sns.set()\n",
    "\n",
    "mf = pymaf.MafFrame.from_file('./data/final.maf')\n",
    "af = common.AnnFrame.from_file('./data/sample-annot.csv', sep=',', sample_col='SampleID')\n",
    "\n",
    "fig, [[ax1, ax2], [ax3, ax4]] = plt.subplots(2, 2, figsize=(15, 15))\n",
    "\n",
    "axes = [ax1, ax2, ax3, ax4]\n",
    "thresholds = [0.25, 0.35, 0.45, 0.55]\n",
    "\n",
    "for i, threshold in enumerate(thresholds):\n",
    "    mf.plot_clonality('AF',\n",
    "                      group_col='Type',\n",
    "                      group_order=['Tissue', 'Baseline', 'PD'],\n",
    "                      af=af, ax=axes[i],\n",
    "                      subclonal=True,\n",
    "                      threshold=threshold)\n",
    "    if i == 0:\n",
    "        axes[i].legend(fontsize=20)\n",
    "    else:\n",
    "        axes[i].get_legend().remove()\n",
    "        \n",
    "    axes[i].set_ylim([0, 1.1])\n",
    "    axes[i].set_title(f'Threshold = {threshold}', fontsize=20)\n",
    "    axes[i].tick_params(axis='both', which='major', labelsize=15)\n",
    "    for ticklabel in axes[i].get_xticklabels():\n",
    "        ticklabel.set_rotation(90)\n",
    "\n",
    "    axes[i].xaxis.label.set_size(20)\n",
    "    axes[i].yaxis.label.set_size(20)\n",
    "    \n",
    "plt.tight_layout()\n",
    "plt.savefig('./outputs/subclone.png')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Supplementary Table 1. Comparison of targeted panels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "genes = {}\n",
    "panels = ['ACP1', 'ACMP', 'SNUBH_V1', 'SNUBH_V2', 'TS500']\n",
    "files = ['acp1.list', 'acmp.list', 'snubh-v1.list', 'snubh-v2.list', 'ts500.list']\n",
    "\n",
    "for i, file in enumerate(files):\n",
    "    n = 0\n",
    "    with open(f'./panels/{file}') as f:\n",
    "        for line in f:\n",
    "            n += 1\n",
    "            gene = line.strip()\n",
    "            if gene not in genes:\n",
    "                genes[gene] = ['X'] * 5\n",
    "            genes[gene][i] = 'O'\n",
    "    print(panels[i], n)\n",
    "\n",
    "df = pd.DataFrame(genes).T\n",
    "df.index.name = 'Gene'\n",
    "df.columns = panels\n",
    "df.to_csv('./outputs/panel-comparison.csv')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
